// matrix-js-sdk/lib/crypto/verification/request/ToDeviceChannel.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/*
Copyright 2018 New Vector Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { randomString } from "../../../randomstring.js";
import { logger } from "../../../logger.js";
import { CANCEL_TYPE, PHASE_STARTED, PHASE_READY, REQUEST_TYPE, READY_TYPE, START_TYPE, VerificationRequest } from "./VerificationRequest.js";
import { errorFromEvent, newUnexpectedMessageError } from "../Error.js";
import { MatrixEvent } from "../../../models/event.js";
/**
 * A key verification channel that sends verification events over to_device messages.
 * Generates its own transaction ids.
 */
export class ToDeviceChannel {
  // userId and devices of user we're about to verify
  constructor(client, userId, devices, transactionId, deviceId) {
    this.client = client;
    this.userId = userId;
    this.devices = devices;
    this.transactionId = transactionId;
    this.deviceId = deviceId;
    _defineProperty(this, "request", void 0);
  }
  isToDevices(devices) {
    if (devices.length === this.devices.length) {
      for (var device of devices) {
        if (!this.devices.includes(device)) {
          return false;
        }
      }
      return true;
    } else {
      return false;
    }
  }
  static getEventType(event) {
    return event.getType();
  }

  /**
   * Extract the transaction id used by a given key verification event, if any
   * @param event - the event
   * @returns the transaction id
   */
  static getTransactionId(event) {
    var content = event.getContent();
    return content && content.transaction_id;
  }

  /**
   * Checks whether the given event type should be allowed to initiate a new VerificationRequest over this channel
   * @param type - the event type to check
   * @returns boolean flag
   */
  static canCreateRequest(type) {
    return type === REQUEST_TYPE || type === START_TYPE;
  }
  canCreateRequest(type) {
    return ToDeviceChannel.canCreateRequest(type);
  }

  /**
   * Checks whether this event is a well-formed key verification event.
   * This only does checks that don't rely on the current state of a potentially already channel
   * so we can prevent channels being created by invalid events.
   * `handleEvent` can do more checks and choose to ignore invalid events.
   * @param event - the event to validate
   * @param client - the client to get the current user and device id from
   * @returns whether the event is valid and should be passed to handleEvent
   */
  static validateEvent(event, client) {
    if (event.isCancelled()) {
      logger.warn("Ignoring flagged verification request from " + event.getSender());
      return false;
    }
    var content = event.getContent();
    if (!content) {
      logger.warn("ToDeviceChannel.validateEvent: invalid: no content");
      return false;
    }
    if (!content.transaction_id) {
      logger.warn("ToDeviceChannel.validateEvent: invalid: no transaction_id");
      return false;
    }
    var type = event.getType();
    if (type === REQUEST_TYPE) {
      if (!Number.isFinite(content.timestamp)) {
        logger.warn("ToDeviceChannel.validateEvent: invalid: no timestamp");
        return false;
      }
      if (event.getSender() === client.getUserId() && content.from_device == client.getDeviceId()) {
        // ignore requests from ourselves, because it doesn't make sense for a
        // device to verify itself
        logger.warn("ToDeviceChannel.validateEvent: invalid: from own device");
        return false;
      }
    }
    return VerificationRequest.validateEvent(type, event, client);
  }

  /**
   * @param event - the event to get the timestamp of
   * @returns the timestamp when the event was sent
   */
  getTimestamp(event) {
    var content = event.getContent();
    return content && content.timestamp;
  }

  /**
   * Changes the state of the channel, request, and verifier in response to a key verification event.
   * @param event - to handle
   * @param request - the request to forward handling to
   * @param isLiveEvent - whether this is an even received through sync or not
   * @returns a promise that resolves when any requests as an answer to the passed-in event are sent.
   */
  handleEvent(event, request) {
    var _arguments = arguments,
      _this = this;
    return _asyncToGenerator(function* () {
      var isLiveEvent = _arguments.length > 2 && _arguments[2] !== undefined ? _arguments[2] : false;
      var type = event.getType();
      var content = event.getContent();
      if (type === REQUEST_TYPE || type === READY_TYPE || type === START_TYPE) {
        if (!_this.transactionId) {
          _this.transactionId = content.transaction_id;
        }
        var _deviceId = content.from_device;
        // adopt deviceId if not set before and valid
        if (!_this.deviceId && _this.devices.includes(_deviceId)) {
          _this.deviceId = _deviceId;
        }
        // if no device id or different from adopted one, cancel with sender
        if (!_this.deviceId || _this.deviceId !== _deviceId) {
          // also check that message came from the device we sent the request to earlier on
          // and do send a cancel message to that device
          // (but don't cancel the request for the device we should be talking to)
          var cancelContent = _this.completeContent(CANCEL_TYPE, errorFromEvent(newUnexpectedMessageError()));
          return _this.sendToDevices(CANCEL_TYPE, cancelContent, [_deviceId]);
        }
      }
      var wasStarted = request.phase === PHASE_STARTED || request.phase === PHASE_READY;
      yield request.handleEvent(event.getType(), event, isLiveEvent, false, false);
      var isStarted = request.phase === PHASE_STARTED || request.phase === PHASE_READY;
      var isAcceptingEvent = type === START_TYPE || type === READY_TYPE;
      // the request has picked a ready or start event, tell the other devices about it
      if (isAcceptingEvent && !wasStarted && isStarted && _this.deviceId) {
        var nonChosenDevices = _this.devices.filter(d => d !== _this.deviceId && d !== _this.client.getDeviceId());
        if (nonChosenDevices.length) {
          var message = _this.completeContent(CANCEL_TYPE, {
            code: "m.accepted",
            reason: "Verification request accepted by another device"
          });
          yield _this.sendToDevices(CANCEL_TYPE, message, nonChosenDevices);
        }
      }
    })();
  }

  /**
   * See {@link InRoomChannel#completedContentFromEvent} for why this is needed.
   * @param event - the received event
   * @returns the content object
   */
  completedContentFromEvent(event) {
    return event.getContent();
  }

  /**
   * Add all the fields to content needed for sending it over this channel.
   * This is public so verification methods (SAS uses this) can get the exact
   * content that will be sent independent of the used channel,
   * as they need to calculate the hash of it.
   * @param type - the event type
   * @param content - the (incomplete) content
   * @returns the complete content, as it will be sent.
   */
  completeContent(type, content) {
    // make a copy
    content = Object.assign({}, content);
    if (this.transactionId) {
      content.transaction_id = this.transactionId;
    }
    if (type === REQUEST_TYPE || type === READY_TYPE || type === START_TYPE) {
      content.from_device = this.client.getDeviceId();
    }
    if (type === REQUEST_TYPE) {
      content.timestamp = Date.now();
    }
    return content;
  }

  /**
   * Send an event over the channel with the content not having gone through `completeContent`.
   * @param type - the event type
   * @param uncompletedContent - the (incomplete) content
   * @returns the promise of the request
   */
  send(type) {
    var uncompletedContent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    // create transaction id when sending request
    if ((type === REQUEST_TYPE || type === START_TYPE) && !this.transactionId) {
      this.transactionId = ToDeviceChannel.makeTransactionId();
    }
    var content = this.completeContent(type, uncompletedContent);
    return this.sendCompleted(type, content);
  }

  /**
   * Send an event over the channel with the content having gone through `completeContent` already.
   * @param type - the event type
   * @returns the promise of the request
   */
  sendCompleted(type, content) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      var result;
      if (type === REQUEST_TYPE || type === CANCEL_TYPE && !_this2.deviceId) {
        result = yield _this2.sendToDevices(type, content, _this2.devices);
      } else {
        result = yield _this2.sendToDevices(type, content, [_this2.deviceId]);
      }
      // the VerificationRequest state machine requires remote echos of the event
      // the client sends itself, so we fake this for to_device messages
      var remoteEchoEvent = new MatrixEvent({
        sender: _this2.client.getUserId(),
        content,
        type
      });
      yield _this2.request.handleEvent(type, remoteEchoEvent, /*isLiveEvent=*/true, /*isRemoteEcho=*/true, /*isSentByUs=*/true);
      return result;
    })();
  }
  sendToDevices(type, content, devices) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      if (devices.length) {
        var deviceMessages = new Map();
        for (var _deviceId2 of devices) {
          deviceMessages.set(_deviceId2, content);
        }
        yield _this3.client.sendToDevice(type, new Map([[_this3.userId, deviceMessages]]));
      }
    })();
  }

  /**
   * Allow Crypto module to create and know the transaction id before the .start event gets sent.
   * @returns the transaction id
   */
  static makeTransactionId() {
    return randomString(32);
  }
}
export class ToDeviceRequests {
  constructor() {
    _defineProperty(this, "requestsByUserId", new Map());
  }
  getRequest(event) {
    return this.getRequestBySenderAndTxnId(event.getSender(), ToDeviceChannel.getTransactionId(event));
  }
  getRequestByChannel(channel) {
    return this.getRequestBySenderAndTxnId(channel.userId, channel.transactionId);
  }
  getRequestBySenderAndTxnId(sender, txnId) {
    var requestsByTxnId = this.requestsByUserId.get(sender);
    if (requestsByTxnId) {
      return requestsByTxnId.get(txnId);
    }
  }
  setRequest(event, request) {
    this.setRequestBySenderAndTxnId(event.getSender(), ToDeviceChannel.getTransactionId(event), request);
  }
  setRequestByChannel(channel, request) {
    this.setRequestBySenderAndTxnId(channel.userId, channel.transactionId, request);
  }
  setRequestBySenderAndTxnId(sender, txnId, request) {
    var requestsByTxnId = this.requestsByUserId.get(sender);
    if (!requestsByTxnId) {
      requestsByTxnId = new Map();
      this.requestsByUserId.set(sender, requestsByTxnId);
    }
    requestsByTxnId.set(txnId, request);
  }
  removeRequest(event) {
    var userId = event.getSender();
    var requestsByTxnId = this.requestsByUserId.get(userId);
    if (requestsByTxnId) {
      requestsByTxnId.delete(ToDeviceChannel.getTransactionId(event));
      if (requestsByTxnId.size === 0) {
        this.requestsByUserId.delete(userId);
      }
    }
  }
  findRequestInProgress(userId, devices) {
    var requestsByTxnId = this.requestsByUserId.get(userId);
    if (requestsByTxnId) {
      for (var request of requestsByTxnId.values()) {
        if (request.pending && request.channel.isToDevices(devices)) {
          return request;
        }
      }
    }
  }
  getRequestsInProgress(userId) {
    var requestsByTxnId = this.requestsByUserId.get(userId);
    if (requestsByTxnId) {
      return Array.from(requestsByTxnId.values()).filter(r => r.pending);
    }
    return [];
  }
}
