// matrix-js-sdk/lib/matrixrtc/CallMembership.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { deepCompare } from "../utils.js";
import { isLivekitFocusActive } from "./LivekitFocus.js";

// Represents an entry in the memberships section of an m.call.member event as it is on the wire

// There are two different data interfaces. One for the Legacy types and one compliant with MSC4143

// MSC4143 (MatrixRTC) session membership data

export var isSessionMembershipData = data => "focus_active" in data;
var checkSessionsMembershipData = (data, errors) => {
  var _data$focus_active;
  var prefix = "Malformed session membership event: ";
  if (typeof data.device_id !== "string") errors.push(prefix + "device_id must be string");
  if (typeof data.call_id !== "string") errors.push(prefix + "call_id must be string");
  if (typeof data.application !== "string") errors.push(prefix + "application must be a string");
  if (typeof ((_data$focus_active = data.focus_active) === null || _data$focus_active === void 0 ? void 0 : _data$focus_active.type) !== "string") errors.push(prefix + "focus_active.type must be a string");
  if (!Array.isArray(data.foci_preferred)) errors.push(prefix + "foci_preferred must be an array");
  // optional parameters
  if (data.created_ts && typeof data.created_ts !== "number") errors.push(prefix + "created_ts must be number");

  // application specific data (we first need to check if they exist)
  if (data.scope && typeof data.scope !== "string") errors.push(prefix + "scope must be string");
  return errors.length === 0;
};

// Legacy session membership data

export var isLegacyCallMembershipData = data => "membershipID" in data;
var checkCallMembershipDataLegacy = (data, errors) => {
  var prefix = "Malformed legacy rtc membership event: ";
  if (!("expires" in data || "expires_ts" in data)) {
    errors.push(prefix + "expires_ts or expires must be present");
  }
  if ("expires" in data) {
    if (typeof data.expires !== "number") {
      errors.push(prefix + "expires must be numeric");
    }
  }
  if ("expires_ts" in data) {
    if (typeof data.expires_ts !== "number") {
      errors.push(prefix + "expires_ts must be numeric");
    }
  }
  if (typeof data.device_id !== "string") errors.push(prefix + "device_id must be string");
  if (typeof data.call_id !== "string") errors.push(prefix + "call_id must be string");
  if (typeof data.application !== "string") errors.push(prefix + "application must be a string");
  if (typeof data.membershipID !== "string") errors.push(prefix + "membershipID must be a string");
  // optional elements
  if (data.created_ts && typeof data.created_ts !== "number") errors.push(prefix + "created_ts must be number");
  // application specific data (we first need to check if they exist)
  if (data.scope && typeof data.scope !== "string") errors.push(prefix + "scope must be string");
  return errors.length === 0;
};
export class CallMembership {
  static equal(a, b) {
    return deepCompare(a.membershipData, b.membershipData);
  }
  constructor(parentEvent, data) {
    this.parentEvent = parentEvent;
    _defineProperty(this, "membershipData", void 0);
    var sessionErrors = [];
    var legacyErrors = [];
    if (!checkSessionsMembershipData(data, sessionErrors) && !checkCallMembershipDataLegacy(data, legacyErrors)) {
      throw Error("unknown CallMembership data. Does not match legacy call.member (".concat(legacyErrors.join(" & "), ") events nor MSC4143 (").concat(sessionErrors.join(" & "), ")"));
    } else {
      this.membershipData = data;
    }
  }
  get sender() {
    return this.parentEvent.getSender();
  }
  get eventId() {
    return this.parentEvent.getId();
  }
  get callId() {
    return this.membershipData.call_id;
  }
  get deviceId() {
    return this.membershipData.device_id;
  }
  get application() {
    return this.membershipData.application;
  }
  get scope() {
    return this.membershipData.scope;
  }
  get membershipID() {
    if (isLegacyCallMembershipData(this.membershipData)) return this.membershipData.membershipID;
    // the createdTs behaves equivalent to the membershipID.
    // we only need the field for the legacy member envents where we needed to update them
    // synapse ignores sending state events if they have the same content.
    else return this.createdTs().toString();
  }
  createdTs() {
    var _this$membershipData$;
    return (_this$membershipData$ = this.membershipData.created_ts) !== null && _this$membershipData$ !== void 0 ? _this$membershipData$ : this.parentEvent.getTs();
  }

  /**
   * Gets the absolute expiry time of the membership if applicable to this membership type.
   * @returns The absolute expiry time of the membership as a unix timestamp in milliseconds or undefined if not applicable
   */
  getAbsoluteExpiry() {
    // if the membership is not a legacy membership, we assume it is MSC4143
    if (!isLegacyCallMembershipData(this.membershipData)) return undefined;
    if ("expires" in this.membershipData) {
      // we know createdTs exists since we already do the isLegacyCallMembershipData check
      return this.createdTs() + this.membershipData.expires;
    } else {
      // We know it exists because we checked for this in the constructor.
      return this.membershipData.expires_ts;
    }
  }

  /**
   * Gets the expiry time of the event, converted into the device's local time.
   * @deprecated This function has been observed returning bad data and is no longer used by MatrixRTC.
   * @returns The local expiry time of the membership as a unix timestamp in milliseconds or undefined if not applicable
   */
  getLocalExpiry() {
    // if the membership is not a legacy membership, we assume it is MSC4143
    if (!isLegacyCallMembershipData(this.membershipData)) return undefined;
    if ("expires" in this.membershipData) {
      // we know createdTs exists since we already do the isLegacyCallMembershipData check
      var relativeCreationTime = this.parentEvent.getTs() - this.createdTs();
      var localCreationTs = this.parentEvent.localTimestamp - relativeCreationTime;
      return localCreationTs + this.membershipData.expires;
    } else {
      // With expires_ts we cannot convert to local time.
      // TODO: Check the server timestamp and compute a diff to local time.
      return this.membershipData.expires_ts;
    }
  }

  /**
   * @returns The number of milliseconds until the membership expires or undefined if applicable
   */
  getMsUntilExpiry() {
    if (isLegacyCallMembershipData(this.membershipData)) {
      // Assume that local clock is sufficiently in sync with other clocks in the distributed system.
      // We used to try and adjust for the local clock being skewed, but there are cases where this is not accurate.
      // The current implementation allows for the local clock to be -infinity to +MatrixRTCSession.MEMBERSHIP_EXPIRY_TIME/2
      return this.getAbsoluteExpiry() - Date.now();
    }

    // Assumed to be MSC4143
    return undefined;
  }

  /**
   * @returns true if the membership has expired, otherwise false
   */
  isExpired() {
    if (isLegacyCallMembershipData(this.membershipData)) return this.getMsUntilExpiry() <= 0;

    // MSC4143 events expire by being updated. So if the event exists, its not expired.
    return false;
  }
  getPreferredFoci() {
    var _this$membershipData$2;
    // To support both, the new and the old MatrixRTC memberships have two cases based
    // on the availablitiy of `foci_preferred`
    if (isLegacyCallMembershipData(this.membershipData)) return (_this$membershipData$2 = this.membershipData.foci_active) !== null && _this$membershipData$2 !== void 0 ? _this$membershipData$2 : [];

    // MSC4143 style membership
    return this.membershipData.foci_preferred;
  }
  getFocusSelection() {
    if (isLegacyCallMembershipData(this.membershipData)) {
      return "oldest_membership";
    } else {
      var focusActive = this.membershipData.focus_active;
      if (isLivekitFocusActive(focusActive)) {
        return focusActive.focus_selection;
      }
    }
  }
}
