// matrix-js-sdk/lib/crypto/RoomList.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/*
Copyright 2018 - 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Manages the list of encrypted rooms
 */

import { IndexedDBCryptoStore } from "./store/indexeddb-crypto-store.js";

/* eslint-disable camelcase */

/* eslint-enable camelcase */

/**
 * Information about the encryption settings of rooms. Loads this information
 * from the supplied crypto store when `init()` is called, and saves it to the
 * crypto store whenever it is updated via `setRoomEncryption()`. Can supply
 * full information about a room's encryption via `getRoomEncryption()`, or just
 * answer whether or not a room has encryption via `isRoomEncrypted`.
 */
export class RoomList {
  constructor(cryptoStore) {
    this.cryptoStore = cryptoStore;
    // Object of roomId -> room e2e info object (body of the m.room.encryption event)
    _defineProperty(this, "roomEncryption", {});
  }
  init() {
    var _this = this;
    return _asyncToGenerator(function* () {
      yield _this.cryptoStore.doTxn("readwrite", [IndexedDBCryptoStore.STORE_ROOMS], txn => {
        _this.cryptoStore.getEndToEndRooms(txn, result => {
          _this.roomEncryption = result;
        });
      });
    })();
  }
  getRoomEncryption(roomId) {
    return this.roomEncryption[roomId] || null;
  }
  isRoomEncrypted(roomId) {
    return Boolean(this.getRoomEncryption(roomId));
  }
  setRoomEncryption(roomId, roomInfo) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      // important that this happens before calling into the store
      // as it prevents the Crypto::setRoomEncryption from calling
      // this twice for consecutive m.room.encryption events
      _this2.roomEncryption[roomId] = roomInfo;
      yield _this2.cryptoStore.doTxn("readwrite", [IndexedDBCryptoStore.STORE_ROOMS], txn => {
        _this2.cryptoStore.storeEndToEndRoom(roomId, roomInfo, txn);
      });
    })();
  }
}
